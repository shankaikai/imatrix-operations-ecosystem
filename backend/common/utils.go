// This file contains useful generic functions that could be used
// in any other packages.

package common

import (
	"fmt"
	"net/http"
	"strings"
	"time"
)

const (
	CUSTOM_NEG_INF  = -999999999999
	DATETIME_FORMAT = "2006-01-02 15:04:05"
	TIME_FORMAT     = "15:04:05"
)

// Checks if an element (key) exists within a sorted list
// Returns if the element is found and index of the key
// in the array if found. If the key is not found, the index
// will be custom -inf for
func BinarySearch(list []int, left int, right int, key int) (bool, int) {
	fmt.Println("Binary Search", left, right, key, list)
	if right < left {
		return false, CUSTOM_NEG_INF
	}

	mid := left + (right-left)/2

	if list[mid] == key {
		return true, mid
	}
	if key < list[mid] {
		return BinarySearch(list, left, mid-1, key)
	} else {
		return BinarySearch(list, mid+1, right, key)
	}
}

func HttpGetWithJWT(url, jwt string) (*http.Response, error) {
	client := &http.Client{
		Timeout: time.Second * 10,
	}
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("HttpGetWithJWT ERROR %s", err.Error())
	}

	bearer := "Bearer " + jwt
	req.Header.Set("Authorization", bearer)
	response, err := client.Do(req)

	if err != nil {
		return nil, fmt.Errorf("HttpGetWithJWT ERROR %s", err.Error())
	}

	return response, nil
}

func HttpPostWithJWT(url, jwt string, body string) (*http.Response, error) {
	client := &http.Client{
		Timeout: time.Second * 10,
	}
	req, err := http.NewRequest("POST", url, strings.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("HttpPostWithJWT ERROR %s", err.Error())
	}

	bearer := "Bearer " + jwt
	req.Header.Set("x-authorization", bearer)
	req.Header.Set("content-type", "application/json")

	response, err := client.Do(req)

	if err != nil {
		return nil, fmt.Errorf("HttpPostWithJWT ERROR %s", err.Error())
	}

	return response, nil
}
