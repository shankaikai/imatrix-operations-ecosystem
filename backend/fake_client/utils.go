package client

import (
	"strconv"
	"time"

	"capstone.operations_ecosystem/backend/common"
	pb "capstone.operations_ecosystem/backend/proto"
	"google.golang.org/protobuf/types/known/timestamppb"
)

func createFakeUser(id int) *pb.User {
	return &pb.User{
		UserId:          int64(id),
		UserType:        pb.User_ISPECIALIST,
		Name:            "test name" + strconv.Itoa(id),
		Email:           "email" + strconv.Itoa(id),
		PhoneNumber:     "1232",
		TelegramHandle:  "sfds",
		UserSecurityImg: "dsfds",
		IsPartTimer:     false,
	}
}

func createFakeClient(id int) *pb.Client {
	return &pb.Client{
		ClientId:     int64(id),
		Name:         "test name" + strconv.Itoa(id),
		Abbreviation: "ABC" + strconv.Itoa(id),
		Email:        "email" + strconv.Itoa(id),
		Address:      "address" + strconv.Itoa(id),
		PostalCode:   int64(123213 + id),
		PhoneNumber:  "1232",
	}
}

func CreateFakeBroadcast(id int, hasRecipient bool) *pb.Broadcast {
	recipients := make([]*pb.BroadcastRecipient, 0)
	aifsRecipients := make([]*pb.AIFSBroadcastRecipient, 0)
	for i := 1; i < 6; i++ {
		recipients = append(recipients, createFakeBroadcastRec(i, hasRecipient))
	}

	for i := 1; i < 3; i++ {
		aifsRecipients = append(aifsRecipients, &pb.AIFSBroadcastRecipient{
			AifsId:    int64(i),
			Recipient: recipients,
		})
	}

	return &pb.Broadcast{
		BroadcastId:  int64(id),
		Type:         pb.Broadcast_ANNOUNCEMENT,
		Content:      "content" + strconv.Itoa(id) + " with some other quotes \"test test 'test'",
		CreationDate: timestamppb.Now(),
		Deadline:     &timestamppb.Timestamp{Seconds: int64(time.Now().Add(30).Unix())},
		Creator:      createFakeUser(1),
		Recipients:   aifsRecipients,
		Urgency:      pb.Broadcast_LOW,
	}
}

func createFakeBroadcastRec(id int, hasUser bool) *pb.BroadcastRecipient {
	var user *pb.User
	if hasUser {
		user = createFakeUser(id)
	}

	return &pb.BroadcastRecipient{
		BroadcastRecipientsId: int64(id),
		Recipient:             user,
		Acknowledged:          false,
		Rejected:              false,
		AifsId:                int64(id),
	}
}

func CreateFakeRoster(id int) *pb.Roster {
	rosterAssignments := make([]*pb.RosterAssignement, 0)
	aifsAssignments := make([]*pb.AIFSClientRoster, 0)

	for i := 2; i < 3; i++ {
		rosterAssignments = append(rosterAssignments, createFakeRosterAssignment(i))
		aifsAssignments = append(aifsAssignments, createFakeAifsAssignment(i))
	}

	startTimeTime := time.Date(2022, 1, 22, 18, 0, 0, 0, time.UTC)
	endTimeTime := time.Date(2022, 1, 22, 6, 0, 0, 0, time.UTC)
	// startTime := &timestamppb.Timestamp{Seconds: startTimeTime.Unix()}
	// endTime := &timestamppb.Timestamp{Seconds: endTimeTime.Unix()}

	return &pb.Roster{
		RosteringId:   int64(id),
		AifsId:        int64(id),
		StartTime:     startTimeTime.Format(common.DATETIME_FORMAT),
		EndTime:       endTimeTime.Format(common.DATETIME_FORMAT),
		Clients:       aifsAssignments,
		GuardAssigned: rosterAssignments,
	}
}

func createFakeRosterAssignment(id int) *pb.RosterAssignement {
	return &pb.RosterAssignement{
		RosterAssignmentId: int64(id),
		GuardAssigned:      createFakeEmployeeEval(id),
		CustomStartTime:    timestamppb.Now(),
		CustomEndTime:      &timestamppb.Timestamp{Seconds: timestamppb.Now().AsTime().Add(time.Hour * 12).Unix()},
		Confirmed:          false,
		Attended:           false,
		AttendanceTime:     nil,
	}
}

func createFakeAifsAssignment(id int) *pb.AIFSClientRoster {
	return &pb.AIFSClientRoster{
		AifsClientRosterId: int64(id),
		Client:             createFakeClient(id),
		PatrolOrder:        int64(id),
	}
}

func createFakeEmployeeEval(id int) *pb.EmployeeEvaluation {
	return &pb.EmployeeEvaluation{
		Employee:      createFakeUser(id),
		EmployeeScore: float32(5 - 0.01*float32(id)),
		IsAvailable:   true,
	}
}

func CreateFakeIncidentReport(id int) *pb.IncidentReport {
	return &pb.IncidentReport{
		IncidentReportId:      int64(id),
		Type:                  pb.IncidentReport_INTRUDER,
		Creator:               createFakeUser(id),
		CreationDate:          time.Now().Format(common.DATETIME_FORMAT),
		LastModifiedDate:      time.Now().Format(common.DATETIME_FORMAT),
		LastModifedUser:       createFakeUser(id),
		IsOriginal:            id%2 == 0,
		IsApproved:            id%2 != 0,
		Signature:             nil,
		ApprovalDate:          "",
		IncidentReportContent: CreateFakeIncidentReportContent(id),
	}
}

func CreateFakeIncidentReportContent(id int) *pb.IncidentReportContent {
	return &pb.IncidentReportContent{
		ReportContentId:       1,
		LastModifiedDate:      time.Now().Format(common.DATETIME_FORMAT),
		LastModifedUser:       createFakeUser(id),
		Address:               "address",
		IncidentTime:          time.Now().Format(common.DATETIME_FORMAT),
		IsPoliceNotified:      false,
		Title:                 "Title",
		Description:           "So there is something here \"a description quotes\" test test",
		HasActionTaken:        false,
		ActionTaken:           "",
		HasInjury:             true,
		InjuryDescription:     "InjuryDescription",
		HasStolenItem:         true,
		StolenItemDescription: "StolenItemDescription",
		ReportImageLink:       "",
	}
}
