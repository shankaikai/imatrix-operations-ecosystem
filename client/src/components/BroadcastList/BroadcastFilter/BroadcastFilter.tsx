import { Button, Group, Select, Text, TextInput } from "@mantine/core";
import React, { Dispatch } from "react";
import { Plus, Search } from "tabler-icons-react";
import { useBroadcastClient } from "../../../helpers/useBroadcastClient";

interface BroadcastFilterProps {
  setModalOpen: Dispatch<boolean>;
}

export default function BroadcastFilter({
  setModalOpen,
}: BroadcastFilterProps) {
  const {
    search,
    setSearch,
    selectValue,
    setSelectValue,
    filterValue,
    setFilterValue,
  } = useBroadcastClient();

  const handleNewClick = () => {
    setModalOpen(true);
  };

  return (
    <Group
      spacing="xs"
      position="apart"
      sx={{
        display: "sticky",
      }}
    >
      {setSearch && (
        <TextInput
          icon={<Search size={16} />}
          placeholder="Search"
          radius="lg"
          size="xs"
          style={{ flex: "1 1 auto" }}
          value={search}
          onChange={(event) => setSearch(event.currentTarget.value)}
        />
      )}

      <Button
        size="xs"
        radius="xl"
        leftIcon={<Plus size={14} />}
        onClick={handleNewClick}
      >
        New
      </Button>
      <Group>
        <Text size="xs" color="dimmed">
          Sort by:
        </Text>
        <Select
          data={[
            { value: "latest", label: "Lastest First" },
            { value: "oldest", label: "Oldest First" },
          ]}
          value={selectValue}
          //@ts-ignore
          onChange={setSelectValue}
          variant="unstyled"
          size="xs"
        />
      </Group>
      <Group>
        <Text size="xs" color="dimmed">
          Filter by:
        </Text>
        <Select
          data={[
            { value: "all", label: "All" },
            { value: "1", label: "AIFS 1" },
            { value: "2", label: "AIFS 2" },
            { value: "3", label: "AIFS 3" },
          ]}
          value={filterValue}
          //@ts-ignore
          onChange={setFilterValue}
          variant="unstyled"
          size="xs"
        />
      </Group>
    </Group>
  );
}
