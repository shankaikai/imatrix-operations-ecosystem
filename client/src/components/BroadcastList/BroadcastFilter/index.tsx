import { Group, TextInput, Select, createStyles, Text } from "@mantine/core";
import { Search } from "tabler-icons-react";
import React, { useState } from "react";

const useStyles = createStyles((theme) => ({
  tabcontainer: {
    height: "100vh",
    width: "360px",
    padding: theme.spacing.xs,
  },
  filters: {
    columnGap: theme.spacing.xs,
  },
  stack: {
    marginTop: theme.spacing.xs,
  },
}));

export default function BroadcastFilter() {
  const [value, setValue] = useState("latest");
  const [search, setSearch] = useState("");

  const { classes } = useStyles();
  return (
    <Group className={classes.filters} position="apart">
      <TextInput
        icon={<Search size={16} />}
        placeholder="Search"
        radius="lg"
        size="xs"
        style={{ width: "170px" }}
        value={search}
        onChange={(event) => setSearch(event.currentTarget.value)}
      />
      <Group>
        <Text size="xs" color="dimmed">
          Sort by:
        </Text>
        <Select
          data={[
            { value: "latest", label: "Lastest First" },
            { value: "oldest", label: "Oldest First" },
          ]}
          value={value}
          //@ts-ignore
          onChange={setValue}
          variant="unstyled"
          size="xs"
          style={{ width: "95px" }}
        />
      </Group>
    </Group>
  );
}
