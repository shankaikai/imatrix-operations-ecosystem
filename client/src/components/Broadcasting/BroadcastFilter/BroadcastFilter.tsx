import { ActionIcon, Button, Group, Select, Text, TextInput } from "@mantine/core";
import React, { Dispatch } from "react";
import { Plus, Search, Refresh } from "tabler-icons-react";
import { useBroadcast } from "../../../helpers/useBroadcastClient";

interface BroadcastFilterProps {
  setModalOpen: Dispatch<boolean>;
}

export default function BroadcastFilter({
  setModalOpen,
}: BroadcastFilterProps) {
  const {
    search,
    setSearch,
    selectValue,
    setSelectValue,
    filterValue,
    setFilterValue,
    updateBroadcasts,
    setBroadcasts
  } = useBroadcast();

  const handleNewClick = () => {
    setModalOpen(true);
  };

  return (
    <Group
      spacing="xs"
      position="apart"
      sx={{
        display: "sticky",
      }}
    >
      {setSearch && (
        <TextInput
          data-testid="broadcastSearch"
          icon={<Search size={16} />}
          placeholder="Search"
          radius="lg"
          size="xs"
          style={{ flex: "1 1 auto" }}
          value={search}
          onChange={(event) => setSearch(event.currentTarget.value)}
        />
      )}

      <Button
        data-testid="newBroadcastButton"
        size="xs"
        radius="xl"
        leftIcon={<Plus size={14} />}
        onClick={handleNewClick}
      >
        New
      </Button>
      <Group>
        <Text size="xs" color="dimmed">
          Sort by:
        </Text>
        <Select
          data-testid="sortBy"
          data={[
            { value: "latest", label: "Latest First" },
            { value: "oldest", label: "Oldest First" },
          ]}
          value={selectValue}
          //@ts-ignore
          onChange={setSelectValue}
          variant="unstyled"
          size="xs"
          sx={{
            width: "100px",
          }}
        />
      </Group>
      <Group>
        <Text size="xs" color="dimmed">
          Filter by:
        </Text>
        <Select
          data-testid="filterBy"
          data={[
            { value: "all", label: "All" },
            { value: "1", label: "AIFS 1" },
            { value: "2", label: "AIFS 2" },
            { value: "3", label: "AIFS 3" },
          ]}
          value={filterValue}
          //@ts-ignore
          onChange={setFilterValue}
          variant="unstyled"
          size="xs"
          sx={{
            width: "100px",
          }}
        />
      </Group>
      <Group>
        <ActionIcon onClick={()=>{(updateBroadcasts && setBroadcasts) && updateBroadcasts(0, setBroadcasts , true)}}>
          <Refresh size={14} />
        </ActionIcon>
      </Group>
    </Group>
  );
}
