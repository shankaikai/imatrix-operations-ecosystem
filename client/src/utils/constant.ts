// Dev
export const ENVOY_ADDRESS = "http://localhost:8080";

// Prod
// export const ENVOY_ADDRESS = "https://operations-envoy.aifs.lunarcloud.org";
