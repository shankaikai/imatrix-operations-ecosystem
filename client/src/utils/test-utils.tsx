import React, { FC, ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import { ColorSchemeProvider, MantineProvider } from "@mantine/core";

const AllTheProviders: FC<{ children: React.ReactNode }> = ({ children }) => {
  const toggleColorScheme = () => {};

  return (
    <ColorSchemeProvider
      colorScheme="light"
      toggleColorScheme={toggleColorScheme}
    >
      <MantineProvider>{children}</MantineProvider>
    </ColorSchemeProvider>
  );
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
