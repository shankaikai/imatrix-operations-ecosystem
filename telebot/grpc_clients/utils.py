# TODO: Create flags

WEB_SERVER_ADDR = "backend"
WEB_SERVER_PORT = 9090

