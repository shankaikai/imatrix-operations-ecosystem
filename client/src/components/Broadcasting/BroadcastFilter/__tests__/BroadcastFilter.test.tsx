import { Dispatch } from "react";
import { render } from "../../../../utils/test-utils";
import BroadcastFilter from "../BroadcastFilter";

const mockDispatch = jest.fn();

describe("BroadcastFilter", () => {
  it("matches snapshot", () => {
    const { container } = render(
      <BroadcastFilter setModalOpen={mockDispatch} />
    );

    expect(container).toMatchSnapshot();
  });
});
