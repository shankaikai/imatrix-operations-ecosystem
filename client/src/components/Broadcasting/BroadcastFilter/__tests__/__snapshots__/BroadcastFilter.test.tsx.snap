// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BroadcastFilter matches snapshot 1`] = `
<div>
  <div
    class="mantine-Group-root mantine-o53c20"
  >
    <button
      class="mantine-Button-filled mantine-Button-root mantine-Group-child mantine-2phkfh"
      data-testid="newBroadcastButton"
      type="button"
    >
      <div
        class="mantine-3xbgk5 mantine-Button-inner"
      >
        <span
          class="mantine-Button-icon mantine-Button-leftIcon mantine-1g4orza"
        >
          <svg
            class="icon icon-tabler icon-tabler-plus"
            fill="none"
            height="14"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="14"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
              stroke="none"
            />
            <line
              x1="12"
              x2="12"
              y1="5"
              y2="19"
            />
            <line
              x1="5"
              x2="19"
              y1="12"
              y2="12"
            />
          </svg>
        </span>
        <span
          class="mantine-qo1k2 mantine-Button-label"
        >
          New
        </span>
      </div>
    </button>
    <div
      class="mantine-Group-root mantine-Group-child mantine-zah578"
    >
      <div
        class="mantine-Text-root mantine-Group-child mantine-1eas2m5"
      >
        Sort by:
      </div>
      <div
        class="mantine-Select-root mantine-Group-child mantine-1k2zk11"
      >
        <div
          aria-controls="mantine-xfwph5ngo"
          aria-expanded="false"
          aria-haspopup="listbox"
          role="combobox"
          tabindex="-1"
        >
          <input
            type="hidden"
            value="latest"
          />
          <div
            class="mantine-Select-wrapper mantine-12sbrde"
          >
            <input
              aria-autocomplete="list"
              aria-invalid="false"
              autocomplete="nope"
              class="mantine-Select-unstyledVariant mantine-Select-input mantine-bxo24b"
              data-mantine-stop-propagation="false"
              data-testid="sortBy"
              id="mantine-xfwph5ngo"
              readonly=""
              type="text"
              value="Latest First"
            />
            <div
              class="mantine-bl0c8b mantine-Select-rightSection"
            >
              <svg
                data-chevron="true"
                fill="none"
                height="14"
                style="color: rgb(134, 142, 150);"
                viewBox="0 0 15 15"
                width="14"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M4.93179 5.43179C4.75605 5.60753 4.75605 5.89245 4.93179 6.06819C5.10753 6.24392 5.39245 6.24392 5.56819 6.06819L7.49999 4.13638L9.43179 6.06819C9.60753 6.24392 9.89245 6.24392 10.0682 6.06819C10.2439 5.89245 10.2439 5.60753 10.0682 5.43179L7.81819 3.18179C7.73379 3.0974 7.61933 3.04999 7.49999 3.04999C7.38064 3.04999 7.26618 3.0974 7.18179 3.18179L4.93179 5.43179ZM10.0682 9.56819C10.2439 9.39245 10.2439 9.10753 10.0682 8.93179C9.89245 8.75606 9.60753 8.75606 9.43179 8.93179L7.49999 10.8636L5.56819 8.93179C5.39245 8.75606 5.10753 8.75606 4.93179 8.93179C4.75605 9.10753 4.75605 9.39245 4.93179 9.56819L7.18179 11.8182C7.35753 11.9939 7.64245 11.9939 7.81819 11.8182L10.0682 9.56819Z"
                  fill="currentColor"
                  fill-rule="evenodd"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="mantine-Group-root mantine-Group-child mantine-zah578"
    >
      <div
        class="mantine-Text-root mantine-Group-child mantine-1eas2m5"
      >
        Filter by:
      </div>
      <div
        class="mantine-Select-root mantine-Group-child mantine-1k2zk11"
      >
        <div
          aria-controls="mantine-cr0ru0ol9"
          aria-expanded="false"
          aria-haspopup="listbox"
          role="combobox"
          tabindex="-1"
        >
          <input
            type="hidden"
            value="all"
          />
          <div
            class="mantine-Select-wrapper mantine-12sbrde"
          >
            <input
              aria-autocomplete="list"
              aria-invalid="false"
              autocomplete="nope"
              class="mantine-Select-unstyledVariant mantine-Select-input mantine-bxo24b"
              data-mantine-stop-propagation="false"
              data-testid="filterBy"
              id="mantine-cr0ru0ol9"
              readonly=""
              type="text"
              value="All"
            />
            <div
              class="mantine-bl0c8b mantine-Select-rightSection"
            >
              <svg
                data-chevron="true"
                fill="none"
                height="14"
                style="color: rgb(134, 142, 150);"
                viewBox="0 0 15 15"
                width="14"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M4.93179 5.43179C4.75605 5.60753 4.75605 5.89245 4.93179 6.06819C5.10753 6.24392 5.39245 6.24392 5.56819 6.06819L7.49999 4.13638L9.43179 6.06819C9.60753 6.24392 9.89245 6.24392 10.0682 6.06819C10.2439 5.89245 10.2439 5.60753 10.0682 5.43179L7.81819 3.18179C7.73379 3.0974 7.61933 3.04999 7.49999 3.04999C7.38064 3.04999 7.26618 3.0974 7.18179 3.18179L4.93179 5.43179ZM10.0682 9.56819C10.2439 9.39245 10.2439 9.10753 10.0682 8.93179C9.89245 8.75606 9.60753 8.75606 9.43179 8.93179L7.49999 10.8636L5.56819 8.93179C5.39245 8.75606 5.10753 8.75606 4.93179 8.93179C4.75605 9.10753 4.75605 9.39245 4.93179 9.56819L7.18179 11.8182C7.35753 11.9939 7.64245 11.9939 7.81819 11.8182L10.0682 9.56819Z"
                  fill="currentColor"
                  fill-rule="evenodd"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="mantine-Group-root mantine-Group-child mantine-zah578"
    >
      <button
        class="mantine-ActionIcon-hover mantine-ActionIcon-root mantine-Group-child mantine-1r9bacs"
        type="button"
      >
        <svg
          class="icon icon-tabler icon-tabler-refresh"
          fill="none"
          height="14"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="14"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
            stroke="none"
          />
          <path
            d="M20 11a8.1 8.1 0 0 0 -15.5 -2m-.5 -4v4h4"
          />
          <path
            d="M4 13a8.1 8.1 0 0 0 15.5 2m.5 4v-4h-4"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
